pkg/auth/oauth/registry/tokentimeouts.go

```diff
@@ -52,56 +54,83 @@ type TokenTimeoutUpdater struct {
 	safetyMargin   time.Duration
 }
 
-func NewTokenTimeoutUpdater(tokens oauthclient.OAuthAccessTokenInterface, oauthClient oauthlister.OAuthClientLister, defaultTimeout int32) *TokenTimeoutUpdater {
+type oauthTokenValidatingAuthenticator struct {
+	delegate  authenticator.OAuthToken
+	validator authenticator.OAuthTokenValidator
+}
+
+func NewValidatingOAuthTokenAuthenticator(delegate authenticator.OAuthToken, validators ...authenticator.OAuthTokenValidator) authenticator.OAuthToken {
+	return &oauthTokenValidatingAuthenticator{
+		delegate:  delegate,
+		validator: authenticator.OAuthTokenValidators(validators),
+	}
+}
+
+func (a *oauthTokenValidatingAuthenticator) AuthenticateOAuthToken(name string) (*oauth.OAuthAccessToken, user.Info, bool, error) {
+	token, user, ok, err := a.delegate.AuthenticateOAuthToken(name)
+	if !ok || err != nil {
+		return token, user, ok, err
+	}
+
+	if err := a.validator.Validate(token); err != nil {
+		return nil, nil, false, err
+	}
+
+	return token, user, ok, err
+}
+
+func NewOAuthTokenTimeoutValidator(tokens oauthclient.OAuthAccessTokenInterface, oauthClient oauthlister.OAuthClientLister, defaultTimeout int32) (authenticator.OAuthTokenValidator, func(stopCh <-chan struct{})) {
 	// flushTimeout is set to one third of defaultTimeout
 	flushTimeout := defaultTimeout / 3
 	if flushTimeout < validation.MinFlushTimeout {
 		flushTimeout = validation.MinFlushTimeout
 	}
 	// safetyMargin is set to one tenth of flushTimeout
 	safetyMargin := flushTimeout / 10
-	ttu := &TokenTimeoutUpdater{
-		oauthClient,
-		tokens,
-		make(chan tokenData),
-		make(map[string]tokenData),
+	ttu := &oauthTokenTimeoutValidator{
+		oauthClient:  oauthClient,
+		tokens:       tokens,
+		tokenChannel: make(chan tokenData),
+		data:         make(map[string]tokenData),
 		// FIXME: what is the right degree for the btree
-		btree.New(32),
-		timeoutAsDuration(defaultTimeout),
-		timeoutAsDuration(flushTimeout),
-		timeoutAsDuration(safetyMargin),
+		tree:           btree.New(32),
+		defaultTimeout: timeoutAsDuration(defaultTimeout),
+		flushTimeout:   timeoutAsDuration(flushTimeout),
+		safetyMargin:   timeoutAsDuration(safetyMargin),
 	}
-	return ttu
-}
-
-func timeoutAsTime(creation time.Time, timeout int32) time.Time {
-	return creation.Add(time.Duration(timeout) * time.Second)
+	return ttu, ttu.start
 }
 
-func timeoutAsDuration(timeout int32) time.Duration {
-	return time.Duration(timeout) * time.Second
-}
-
-// CheckTimeout is called with a token when it is seen by an authenticator
+// Validate is called with a token when it is seen by an authenticator
 // it touches only the tokenChannel so it is safe to call from other threads
-func (a *TokenTimeoutUpdater) CheckTimeout(name string, token *oauth.OAuthAccessToken) error {
+func (a *oauthTokenTimeoutValidator) Validate(token *oauth.OAuthAccessToken) error {
+	if token.TimeoutsIn == 0 {
```

Comment:
This changes the behavior, adding a timeout on a oauthclient does not cause existing tokens to start having a timeout, is this intentional ?

Reply:
Fixed.